### ISSUE TITLE
Theme support: more flexible custom class names and support for inline styles

### ISSUE THREAD

**Issue by kof** [bvaughn/react-virtualized#197]
Can we make global class names optional? There are a number of ways, for e.g. passing a theme object like react-theamable suggests.

I am using [jss](https://github.com/jsstyles/jss) and would like to have consistency in class names.

**Comment by bvaughn** [comment]
Hey @kof,

Styling is kind of a difficult issue for libraries like RV because I don't think the community has really embraced a standard way of doing things. I don't think it's possible for RV to support everyone's styling approach (because they're too varied) so I tried to pick one that's the most flexible and open. A good bit of thought and discussion went into it (see issues like #48). The benefits are:
* Supports universal applications (ie. works well for browser and Node, something an earlier version of inline styles did not do due to its auto-prefixer).
* Customizable on a per-element basis as well as entirely (via a fork of the stylesheet).
* Doesn't add much clutter or complexity to RV library code.

That being said I realize it also has some downsides. I'd like to avoid using global class names because of their potential for conflict. I'd also (personally) prefer to use inline styles so that users didn't have to remember to include a stylesheet during their bootstrapping process. (People forget to do this pretty frequently.) But inline styles have priority over external styles and so I avoid them because it would make customization more difficult for RV users who didn't also embrace inline styles. (As I mentioned above it has also caused some complications WRT auto-prefixing in the past.)

RV also allows you to supply your own class names for elements (to be appended to the classnames list) so this should hopefully allow you to use your own approach to styling for customization purposes at least somewhat?

Happy to discuss this more.

**Comment by kof** [comment]
I think there is a universal way which will work for every one: allow components to accept a `theme` object which is a map of names and class names:
```javascript
const theme = {
   button: 'my-button'
}
<List theme={theme} />
```

This allows to pass default global class names by default as it is right now and at the same time overwrite  everything by a user.

**Comment by kof** [comment]
> RV also allows you to supply your own class names for elements 
yes, but not to all of them and global classes are still a problem then.

**Comment by bvaughn** [comment]
Sure, that's true. I only added custom `className` support for what I saw as the biggest value components to avoid props-explosion.

You say global classes are "a problem" but... why? If you're using JSS or CSS-modules their presence shouldn't affect you since both allow you to namespace your application styles.

I don't want to go the route of react-theamable in this case. I think it works well for a whitelist of styles (eg. specific colors or font sizes you want to make customizable) but not for all of the styles a library like react-virtualized needs (eg. flex, overflow, etc).

Your suggestion of a parameter that specifies `className`s is potentially something I could do. But I think it has a couple of potential downsides (though minor):
* I wouldn't want to support a _third_ option for customizing styles so this sort of change would probably have to wait until the next major release version.
* It seems ambiguous what should happen if a property (class name) is missing from this param. Should RV fallback to its default class name? Should it not attach any class names at all to the element in question?
* I don't want to make simple customizations more difficult. For example, I think it's currently common for users to want to customize a single attribute or a grid (eg. background color, border style) but otherwise stick with the default styles. With the approach you describe, I think users would have specify a lot of boilerplate names just to override a specific style. Granted there are ways to reduce that boilerplate but it seems like something the current approach (additive class names) avoids nicely.

**Comment by kof** [comment]
> You say global classes are "a problem" but... why? If you're using JSS or CSS-modules their presence shouldn't affect you since both allow you to namespace your application styles.

Imagine the a case where external components or widgets are used at the same time. You can never know what class names they use. Also it's just not nice to have them around when there is an easy way to get rid of them.

Regarding breaking changes: I totally see we can do it very nice and without breaking changes.

Theme is an object with a given interface which can be validated. If user passes a theme object he will pass all required props. If not - also fine, global names as now are used and we expect the style sheet loaded to the document by user. In those cases where user can pass a class name - nothing changes, theme is theme, it is global, className is something that one might want to change at runtime. No conflicts there and also everything is good to have.

**Comment by bvaughn** [comment]
I understand that there's potential for conflicting class names. That's a limitation of static class names, yup. On the other hand, it's already broadly in use and (up until this point) I haven't heard of any conflicts. So it seems a little bit more like a conceptual problem than a practical one.

I don't mean to write off your observation that it's not ideal. I love CSS modules for what it's worth because they avoid this problem. I guess I'm just pointing out that... in my opinion... the potential downside of static class names needs to be weighed against:

1. The convenience for users when it comes to overriding one-off styles and...
1. My personal time when it comes to supporting a variety of different ways to customize and style a component; (more ways means more possibility bugs and more maintenance effort).

I'm afraid I don't understand your proposal. It was a long day (at work) so maybe I'm just mis-reading it but could you try to restate what you're suggesting?

**Comment by kof** [comment]
Sorry, its probably my bad in explaining those ideas. I will prototype something so that you can see it in code.

**Comment by kof** [comment]
I have created a pull request on another project regarding theming, we can do here exactly the same https://github.com/elementalui/elemental/pull/150/files

**Comment by bvaughn** [comment]
Hi @kof.

Thanks for the update. Read your PR and the discussion attached, and the spec you linked to. It all looks pretty solid.

I have a concern for react-virtualized usage though. I mentioned it [before](https://github.com/bvaughn/react-virtualized/issues/197#issuecomment-211986300) but it will be easier to talk about now that I can see code. With the PR you linked to, I think it would be more work to do small style modifications (eg. a custom background color or border) because a user would need to duplicate an entire `theme` object (including all of the classes/states they don't care about customizing).

I think this concern could be reduced by exposing some sort of theme-creation factory function. I [added some thoughts](https://github.com/elementalui/elemental/pull/150#issuecomment-213989754) on the other PR for one way we could address that concern.

**Comment by bvaughn** [comment]
I realize this morning that it's perhaps a little ambiguous to anyone seeing only the discussion on this issue- but I've decided to move forward with the theming approach we're discussing (or at least one similar) as part of the version-7 update that I'm currently working on.

**Comment by kof** [comment]
Can we collaborate on [theme-standard](https://github.com/kof/theme-standard) at the same time? Means adding docs there and defining the workflow for theming in general.

**Comment by bvaughn** [comment]
@kof I'm not sure if the idea of supporting "themeable" things like colors, fonts, etc. really applies to react-virtualized. I was considering for inspiration something like [alexkuz/react-json-tree](https://github.com/alexkuz/react-json-tree#theming) minus the base16 theme bit since react-virtualized doesn't really have colors.

What did you have in mind in terms of collaboration though? I'm a fan of collaborating in general.

PS. When can I see more of your dynamic height spike? ;)

**Comment by kof** [comment]
Will ping you on gitter.

**Comment by bvaughn** [comment]
I've decided not to move forward with a theme object after much consideration. I _have_ however added inline style support to the upcoming version 7 release (currently available on NPM as version `7.0.0-rc.1`). I appreciate all of the thoughtful discussion and ideas.

It was a difficult decision, but I feel the existing interface for specifying custom classnames for react-virtualized components is simple and concise. The proposed theme object would have been more verbose to use and I didn't feel the benefits outweighed the cost for users.

There were also some edge-cases I don't think fit nicely into the theme approach- for example, a callback that returns the classname for a row given its index. I could have supported such a callback as a one-off but I'd rather not mix and match styling approaches. I believe consistency is preferable (and also easier for me to support).
